### PR TITLE
Add execution timing to cells

### PR DESCRIPTION
**Pull request by Madhu94**
Closes https://github.com/jupyterlab/jupyterlab/issues/3320

Moves forward @saulshanabrook's work in https://github.com/jupyterlab/jupyterlab/pull/5009

Changes from earlier PR - rebase on master, fix tests and movie 'timing' into its own subdict. There was also a bug which prevented the iopub hook of outputarea from being invoked in case recordTiming was on - [here](https://github.com/jupyterlab/jupyterlab/pull/5009/files#diff-901b6622e3beb3db23ab3289d24f5583R810) was where it was overridden.

I had a query in https://github.com/jupyterlab/jupyterlab/pull/5009#issuecomment-513493860, I chose to not use a nested dict, I can change it if people feel otherwise

Can we also consider making the command to record timings sticky? Through the settings registry. 
I believe there is a plan to make this a pattern for all commands? This PR - https://github.com/jupyterlab/jupyterlab/pull/5659

### CODE REVIEW

File: packages/cells/src/widget.ts [hunk 1]
```diff
@@ -1037,10 +1035,52 @@ export namespace CodeCell {
...
+              break;
+            case 'execute_input':
+              label = 'execute_input';
+              break;
+          }
+          const value = msg.header.date;
+          if (!value) {
+            return;
+          }
+          if (!model.metadata.has('timing')) {
```
Comment (saulshanabrook) [inline packages/cells/src/widget.ts]: I think these next two lines could be simplified, like:

```typescript
const timingInfo = model.metadata.get('timing') || {}
timingInfo[`iopub.${label}`] = value
model.metadata.set('timing', timingInfo);
```

**Comment by saulshanabrook** [comment]
> I had a query in #5009 (comment), I chose to not use a nested dict, I can change it if people feel otherwise

That's fine with me. Could you give a listing of all the possible keys of the timing metadata now?

It would be good to also open a PR against the nbformat repo to suggest this change to the official notebook spec. The idea is that the timing object would be a mapping of arbitrary keys to date strings. The question of "How long did this cell take to execute?" would not be directly answerable by these keys. Different UIs would be free to display different keys as they see fit. They represent different semantics of the timing question.

**Comment by Madhu94** [comment]
> The idea is that the timing object would be a mapping of arbitrary keys to date strings. The keys represent event names. 

Like [this](https://github.com/Madhu94/nbformat/commit/9e01a66b7cc38f5dacb27751bcdbdcd8a1445351#diff-b4de71ae1a0a055b38eab43f1ac9876c)?

**Comment by saulshanabrook** [comment]
> Like this?

That looks good! I might make the type a little more specific, so that we specify the keys and values are both strings.

**Comment by Madhu94** [comment]
Also a sample of the timing dict's keys - 

iopub.execute_input
iopub.status.busy
shell.execute_reply.started
shell.execute_reply
iopub.status.idle

I created https://github.com/jupyter/nbformat/pull/144 for the nbformat changes.

**Comment by Madhu94** [comment]
@saulshanabrook Do you think we should consider [this suggestion](https://github.com/jupyterlab/jupyterlab/pull/5009#issuecomment-517026572) ? Seems like a good idea to me.

**Comment by saulshanabrook** [comment]
For the record, he suggested:

> Note that we always record (and display in a hover) which kernel was used in evaluating the cell. Is there something in the spec (or envisioned for it) that records this? People can change the kernel at any time while going through and evaluating cells...


@Madhu94  This does make some sense, but I would like to cover this in a separate issue, since I think this can be done without that feature.

**Comment by Madhu94** [comment]
Okay, then let's proceed with review/merge ?

**Comment by saulshanabrook** [comment]
I am going to bring this up at [this week's dev meeting](https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting). Feel free to attend if you would like!

**Comment by Madhu94** [comment]
> I am going to bring this up at [this week's dev meeting](https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting). Feel free to attend if you would like!

I couldn't join. Or at least no one was there when I connected. I assume it should happen about now.

**Comment by saulshanabrook** [comment]
@Madhu94 it should be tomorrow

**Comment by Madhu94** [comment]
Oh. I got the timezone wrong then (I did use the helpful app that was linked :) )  Wed 9 AM PST seemed to point to Tue 9.30 PM IST.

**Comment by saulshanabrook** [comment]
Sounds good, yeah it will be 24 hours after than when you posted your comment today. If you do not attend, I will also post notes here on what is discussed. I wanted to bring up the question of  storing whether to record execute timing in the notebook metadata makes sense.

**Comment by Madhu94** [comment]
From the dev meeting -
- We decided to rename the 'timing' key to 'execution' and experiment with this first, which I think means I should suspend the nbformat PR for a bit.

I'll also take a look at the existing nbextension and see how easy it would be to make a UI for displaying this (even if the UI is done in userland)

**Comment by blink1073** [comment]
FYI @Madhu94 I just gave you write permissions to this repo.  You should be able to set labels and merge PRs.   Thanks for all your help!

**Comment by saulshanabrook** [comment]
@Madhu94 Since you renamed the `timing` key to `execution`, it should also be renamed on the `nbformat` PR with a description of the intent (i.e. there might be other key, value pairs now that are not timing information, like the kernel name)

**Comment by Madhu94** [comment]
Thanks for reminding me! I meant to _suspend_ the nbformat PR since we planned to leave it as an experimental feature in lab. Let me know if that was misunderstood?

**Comment by Madhu94** [comment]
Yeah, makes sense. I'll update the nbformat PR instead of closing it.

**Comment by Madhu94** [comment]
Taken care ^

**Comment by mlucool** [comment]
@blink1073 @saulshanabrook Is there any reason not to merge this now?

**Comment by saulshanabrook** [comment]
Thank you @Madhu94 for keeping on this for so long, despite the slow review on my  part. I am  going to merge this in. This enables third  party extension to create UIs which show the timing per cell  and for us to experiment with  different ways of displaying this.

**Comment by Madhu94** [comment]
This was quick, thanks for the help:)

Also, I started attempting a UI extension for this but adding a footer to a cell widget in lab seems unnaturally hard (compared to how easy it is to extend other parts of lab)

**Comment by saulshanabrook** [comment]
> I started attempting a UI extension for this but adding a footer to a cell widget in lab seems unnaturally hard (compared to how easy it is to extend other parts of lab)

Ah yeah, I believe that! We are in the process  of refactoring the cell creation widgets anyways for the RTC work... I think though if  you found out a place in  core to add an extension point for a post-cell widget, we could add that as an extension point. Or just some way generally to process a cell before it is displayed. 

Might be a good thing to bring up at this weeks meeting if you wanna attend.

**Comment by jakirkham** [comment]
Is there an issue to track exposing this information in the UI?

**Comment by Madhu94** [comment]
@jakirkham It is unclear at this point if this should be part of Lab or an extension. Note that this wasn't part of the Notebook either, it was just an extension

**Comment by jakirkham** [comment]
Sure that makes sense. Am just curious if there is an issue to subscribe to about updates on the UI portion. Sounds like not ATM, which is fine.

**Pull request merged**